### DISCUSSION TITLE
Accessing theme using composition

### DISCUSSION THREAD

**Discussion by flugg** [emotion-js/emotion#2405] [Q&A]
Hi! If we're making reusable CSS objects as a callback to access the theme:

```ts
const textColor = (theme: Theme) => css({
  color: theme.colors.primary
});

const background = (theme: Theme) => css({
  background: theme.colors.secondary
});
```

We can use these like this:

```ts
<div css={textColor}></div>
```
```ts
<div css={background}></div>
```

However, if we want to apply both of them, this throws a TypeScript error:

```ts
<div css={[textColor, background]}></div>
```

Is this intentional? It feels a bit verbose to do the following:

```ts
<div css={(theme: Theme) => [textColor(theme), background(theme)]}></div>
```

Also, a bonus question: why do we have to type `Theme` manually when using the `css` prop? TypeScript complains about the parameter being implicitly `any`.

**Comment by Andarist** [comment]
> However, if we want to apply both of them, this throws a TypeScript error:

I don't think this should result in an error because support for this has been implemented explicitly as a new feature for v11. Could you create a codesnadbox demonstrating the problem?

> 
<div class="TimelineItem pt-0 js-comment-container discussion-timeline-item ml-0" data-gid="MDEwOkRpc2N1c3Npb24zNDE0ODE2" data-url="/emotion-js/emotion/discussions/2405/body" style="box-sizing: border-box; margin-left: 16px; padding: 16px 0px; position: relative; display: flex; color: rgb(201, 209, 217); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="discussions-timeline-scroll-target js-targetable-element" id="discussion-3414816" style="box-sizing: border-box; width: 904px; padding-top: 60px; margin-top: -60px; pointer-events: none !important;"><div class="
      timeline-comment-group
      reactions-with-gap
      discussion-comment
      js-discussion-comment
      js-minimizable-comment-group
      js-targetable-element
      TimelineItem-body
      js-timeline-discussion-comment my-1
    " data-error="" style="box-sizing: border-box; margin-top: 4px; margin-bottom: 4px !important; min-width: 0px; max-width: 100%; color: var(--color-timeline-text); flex: 1 1 auto; pointer-events: auto;"><div class="d-flex" style="box-sizing: border-box; display: flex !important;"><div class="width-full" style="box-sizing: border-box; width: 904px;"><div class="timeline-comment unminimized-comment comment previewable-edit js-task-list-container flex-auto box-shadow-none js-comment reorderable-task-lists" data-body-version="c306e661b8070af3dfbe1d7721d592a6995c3b6d343345a199c2b61ec0c57a40" data-unfurl-hide-url="/content_reference_attachments/hide" style="box-sizing: border-box; box-shadow: none !important; flex: 1 1 auto !important; position: relative; color: var(--color-text-primary); background-color: var(--color-bg-primary); border: 1px solid var(--color-border-primary); border-radius: 6px;"><div class="" style="box-sizing: border-box;"><div class="edit-comment-hide" style="box-sizing: border-box;"><task-lists disabled="" sortable="" style="box-sizing: border-box;">Also, a bonus question: why do we have to type&nbsp;Theme&nbsp;manually when using the&nbsp;css&nbsp;prop? TypeScript complains about the parameter being implicitly&nbsp;any.</task-lists><div class="d-flex flex-justify-between flex-content-center" style="box-sizing: border-box; place-content: center space-between !important; display: flex !important;"><div class="comment-reactions d-flex flex-items-start flex-row-reverse js-reactions-container js-reaction-buttons-container flex-justify-end" style="box-sizing: border-box; flex-direction: row-reverse !important; justify-content: flex-end !important; align-items: flex-start !important; display: none; margin-left: 16px; border-top: 0px !important;"><form class="js-pick-reaction" action="https://github.com/emotion-js/emotion/reactions" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"><div class="comment-reactions-options js-comment-reactions-options" style="box-sizing: border-box;"></div></form><details class="details-overlay details-reset dropdown hx_dropdown-fullscreen position-relative d-inline-block reaction-popover-container reactions-menu js-reaction-popover-container js-comment-header-reaction-button" style="box-sizing: border-box; display: inline-block !important; position: relative;"><summary class="add-reaction-btn btn-sm Link--secondary discussion-add-reaction-button mb-3 mr-3 color-bg-tertiary border color-border-primary rounded-2" aria-label="Add your reaction" aria-haspopup="menu" role="button" style="box-sizing: border-box; display: list-item; cursor: pointer; padding: 3px 12px; font-size: 12px; line-height: 20px; color: var(--color-text-secondary) !important; border-top-color:  !important; border-top-style:  !important; border-top-width:  !important; border-right: 0px; border-bottom-color:  !important; border-bottom-style:  !important; border-bottom-width:  !important; border-left-color:  !important; border-left-style:  !important; border-left-width:  !important; border-image-source:  !important; border-image-slice:  !important; border-image-width:  !important; border-image-outset:  !important; border-image-repeat:  !important; border-radius: 6px !important; background-color: var(--color-bg-tertiary) !important; margin-bottom: 16px !important; margin-right: 16px !important; opacity: 1; transition: opacity 0.1s ease-in-out 0s; list-style: none;"><svg aria-hidden="true" viewBox="0 0 16 16" version="1.1" data-view-component="true" height="18" width="18" class="octicon octicon-smiley"><path fill-rule="evenodd" d="M1.5 8a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0zM8 0a8 8 0 100 16A8 8 0 008 0zM5 8a1 1 0 100-2 1 1 0 000 2zm7-1a1 1 0 11-2 0 1 1 0 012 0zM5.32 9.636a.75.75 0 011.038.175l.007.009c.103.118.22.222.35.31.264.178.683.37 1.285.37.602 0 1.02-.192 1.285-.371.13-.088.247-.192.35-.31l.007-.008a.75.75 0 111.222.87l-.614-.431c.614.43.614.431.613.431v.001l-.001.002-.002.003-.005.007-.014.019a1.984 1.984 0 01-.184.213c-.16.166-.338.316-.53.445-.63.418-1.37.638-2.127.629-.946 0-1.652-.308-2.126-.63a3.32 3.32 0 01-.715-.657l-.014-.02-.005-.006-.002-.003v-.002h-.001l.613-.432-.614.43a.75.75 0 01.183-1.044h.001z"></path></svg></summary></details></div></div></div><form class="js-comment-update" action="https://github.com/emotion-js/emotion/discussions/2405" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"><div class="js-edit-comment-form-container" data-body-version="c306e661b8070af3dfbe1d7721d592a6995c3b6d343345a199c2b61ec0c57a40" data-outer-div-classes="" data-permissions-url="/emotion-js/emotion/discussions/2405/permissions" data-repository-id="92570536" data-target-body="Hi! If we're making reusable CSS objects as a callback to access the theme:

```ts
const textColor = (theme: Theme) => css({
  color: theme.colors.primary
});

const background = (theme: Theme) => css({
  background: theme.colors.secondary
});
```

We can use these like this:

```ts
<div css={textColor}></div>
```
```ts
<div css={background}></div>
```

However, if we want to apply both of them, this throws a TypeScript error:

```ts
<div css={[textColor, background]}></div>
```

Is this intentional? It feels a bit verbose to do the following:

```ts
<div css={(theme: Theme) => [textColor(theme), background(theme)]}></div>
```

> Also, a bonus question: why do we have to type Theme manually when using the css prop? TypeScript complains about the parameter being implicitly any.

Do you refer to functions like this?
```ts
const textColor = (theme: Theme) => css({
  color: theme.colors.primary
});
```

It's really more of a TS question than Emotion's one - this is how TS works, this function doesn't have enough "context" for TS to infer its parameter types. It's a standalone function and as such all of its parameters have to be annotated explicitly, the same would apply here:
```ts
const add = (a, b) => a + b
```
In here TS can't assess on its own that `a` and `b` should be `number`s (or `string`s)

**Reply by flugg** [reply]
Thanks for the quick reply @Andarist!

Very strange regarding passing an array of callbacks to the `css` prop. Everything else works so I don't think there's anything wrong with our setup. I get the following TypeScript error:

```ts
Type '((theme: Theme) => SerializedStyles)[]' is not assignable to type '((string | number | boolean | ComponentSelector | Keyframes | SerializedStyles | ArrayInterpolation<undefined> | ObjectInterpolation<...> | ((theme: any) => Interpolation<...>)) & (string | ... 7 more ... | FunctionInterpolation<...>)) | null | undefined'.
  Type '((theme: Theme) => SerializedStyles)[]' is not assignable to type '((theme: any) => Interpolation<undefined>) & ArrayInterpolation<Theme>'.
    Type '((theme: Theme) => SerializedStyles)[]' is not assignable to type '(theme: any) => Interpolation<undefined>'.
      Type '((theme: Theme) => SerializedStyles)[]' provides no match for the signature '(theme: any): Interpolation<undefined>'.ts(2322)
index.d.ts(84, 5): The expected type comes from property 'css' which is declared here on type 'ClassAttributes<HTMLParagraphElement> & HTMLAttributes<HTMLParagraphElement> & { css?: Interpolation<Theme>; }'
(property) css?: ((string | number | boolean | ComponentSelector | Keyframes | SerializedStyles | ArrayInterpolation<undefined> | ObjectInterpolation<...> | ((theme: any) => Interpolation<...>)) & (string | ... 7 more ... | FunctionInterpolation<...>)) | null | undefined
```

Not sure if that's helpful. I'll see if I can reproduce the problem in a sandbox tomorrow.

Regarding the bonus question, I'm sorry for not making my question clear enough, I wasn't talking about the TS arrow functions, but rather about accessing theme with the `css` prop. This fails with implicit `any`:

```tsx
<div css={theme => ({})} />
```

And we have to typehint `Theme` manually everywhere like this:

```tsx
<div css={(theme: Theme) => ({})} />
```

**Reply by Andarist** [reply]
> And we have to typehint Theme manually everywhere like this:

This shouldn't be required if you extend our `Theme` correctly (like described in the docs) - I've recently done that in one of the projects at my company and it works like a charm. If you provide a repro case I could try to help you find a problem.

**Reply by flugg** [reply]
I didn't manage to replicate this in a Codesandbox - but after a couple hours of debugging, it turns out a typing file from [Sayegh7/emotion-reset](https://github.com/Sayegh7/emotion-reset) messed up the Emotion types.

I've added a PR which hopefully fixes this: https://github.com/Sayegh7/emotion-reset/pull/64

Sorry for wasting your time on this!
